test/dlc_tests/ec2/test_smdebug.py

```diff
@@ -10,73 +10,70 @@
+    )
 
 
 @pytest.mark.flaky(reruns=0)
 @pytest.mark.integration("smdebug")
 @pytest.mark.model("mnist")
 @pytest.mark.parametrize("ec2_instance_type", SMDEBUG_EC2_CPU_INSTANCE_TYPE, indirect=True)
-def test_smdebug_cpu(training, ec2_connection, region, cpu_only, py3_only):
-    # TODO: Remove this once test timeout has been debugged (failures especially on m4.16xlarge)
-    if is_tf2(training) and "m4.16xlarge" in SMDEBUG_EC2_CPU_INSTANCE_TYPE:
-        pytest.skip("Currently skipping for TF2 on m4.16xlarge until the issue is fixed")
-    if is_tf1(training):
-        pytest.skip("Currently skipping for TF1 until the issue is fixed")
-    run_smdebug_test(training, ec2_connection, region)
+def test_smdebug_cpu(training, ec2_connection, region, ec2_instance_type, cpu_only, py3_only):
+    run_smdebug_test(training, ec2_connection, region, ec2_instance_type)
+
+
+class SMDebugTestFailure(Exception):
+    pass
 
 
 def run_smdebug_test(
     image_uri,
     ec2_connection,
     region,
+    ec2_instance_type,
     docker_executable="docker",
     container_name="smdebug",
     test_script=SMDEBUG_SCRIPT,
-    logfile="output.log",
-    is_large_shm=False
 ):
+    large_shm_instance_types = ("p2.8xlarge", "m4.16xlarge")
+    shm_setting = " --shm-size=1g " if ec2_instance_type in large_shm_instance_types else " "
     framework = get_framework_from_image_uri(image_uri)
     container_test_local_dir = os.path.join("$HOME", "container_tests")
     ec2_connection.run(f"$(aws ecr get-login --no-include-email --region {region})", hide=True)
 
-    shm_setting = '--shm-size=1g' if is_large_shm else ""
-    ec2_connection.run(
-        f"{docker_executable} run {shm_setting} --name {container_name} -v "
-        f"{container_test_local_dir}:{os.path.join(os.sep, 'test')} -itd {image_uri}",
-        hide=True,
-    )
-
     try:
-        test_output = ec2_connection.run(
-            f"{docker_executable} exec --user root {container_name} "
-            f"/bin/bash -c '{test_script} {framework}' | tee {logfile}",
+        ec2_connection.run(
+            f"{docker_executable} run --name {container_name} -v "
+            f"{container_test_local_dir}:{os.path.join(os.sep, 'test')}{shm_setting}{image_uri} "
+            f"./{test_script} {framework}",
             hide=True,
-            warn=True,
-            timeout=3000,
+            timeout=2400,
         )
-    except Exception:
-        debug_output = ec2_connection.run(f"cat {logfile}")
-        LOGGER.error(f"Caught exception while trying to run test via fabric. Output: {debug_output.stdout}")
-        raise
-
-    # LOGGER.info(test_output.stdout)  # Uncomment this line for a complete log dump
-
-    assert test_output.ok, f"SMDebug tests failed. Output:\n{test_output.stdout}"
+    except Exception as e:
+        debug_output = ec2_connection.run(f"docker logs {container_name}")
+        debug_stdout = debug_output.stdout
+        if "All SMDebug tests succeeded!" in debug_stdout:
```

Comment:
In some instance types, fabric connection times out even though the command continues to run on the container. We can check the docker logs on the container itself to see if the test completed successfully - in this case, we will warn, but not fail the test, since SMDebug test passed.

A potential future solution to this is to break SMDebug tests into several smaller tests, and add more regular logging to the tests. In cases of timeout, fabric times out waiting for the next "." to be populated from pytest. However, since we do not control the tests in the sagemaker-debugger repo, this workaround allows us to account for false negatives.

I will follow up in a future PR to test breaking SMDebug tests into smaller parts, but that is out of scope for this fix.

Reply:
What does
> fabric times out waiting for the next "." to be populated from pytest

mean?

Since we don't run pytest commands inside test scripts on any other tests apart from this, is this a unique error that occurs only in smdebug tests?

Reply:
When pytest runs, it will populate a "." to the terminal when a test is completed. We could provide more logging here and see if that helps, but that will **intensely** flood logs when there is a failure, making it especially difficult to identify failures.

Yes, this is unique to smdebug tests because we are not only running pytest on an ec2 instance via a shell script thru fabric, but for longer running tests (like the keras tests), there is a long wait time between shell updates.